File: analyzers/tests/SonarAnalyzer.Test/TestCases/MarkWindowsFormsMainWithStaThread.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Tests.Diagnostics
{
    class Program_00
    {
        [SomeNonsense] // Error [CS0246,CS0246] - unknown type
        // Error@+2 [CS0017] Program has more than one entry point
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static void Main() // Noncompliant {{Add the 'STAThread' attribute to this entry point.}}
        {
        }
    }

    class Program_01
    {
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static void Main() // Noncompliant {{Add the 'STAThread' attribute to this entry point.}}
//                         ^^^^
        {
        }
    }

    class Program_02
    {
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static int Main(string[] args) // Noncompliant {{Add the 'STAThread' attribute to this entry point.}}
//                        ^^^^
        {
            return 1;
        }
    }

    class Program_03
    {
        [MTAThread]
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static void Main() // Noncompliant {{Change the 'MTAThread' attribute of this entry point to 'STAThread'.}}
        {
        }
    }

    class Program_04
    {
        [MTAThread]
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static int Main(string[] args) // Noncompliant {{Change the 'MTAThread' attribute of this entry point to 'STAThread'.}}
        {
            return 1;
        }
    }

    class Program_05
    {
        [System.MTAThreadAttribute]
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static int Main(string[] args) // Noncompliant {{Change the 'MTAThread' attribute of this entry point to 'STAThread'.}}
        {
            return 1;
        }
    }

    class Program_06
    {
        [STAThread]
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static void Main()
        {
        }
    }

    class Program_07
    {
        [STAThread]
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static int Main(string[] args)
        {
            return 1;
        }
    }

    class Program_08
    {
        [System.STAThread]
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static int Main(string[] args)
        {
            return 1;
        }
    }

    class Program_09
    {
        [STAThread("this is wrong", 1)] // Error [CS1729] - ctor doesn't exist
        // Secondary@+1 [CS0017] Program has more than one entry point
        public static int Main(string[] args)
        {
            return 1;
        }
    }

    class Program_10
    {
        public static async Task Main() // Compliant, async Main is always MTA
        {
            await Task.CompletedTask;
        }
    }

    class Program_11
    {
        [STAThread]
        public static async Task Main()
        {
            await Task.CompletedTask;
        }
    }

    class Program_12
    {
        [MTAThread]
        public static async Task Main()
        {
            await Task.CompletedTask;
        }
    }

}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/MarkWindowsFormsMainWithStaThread.vb
' Error [BC30738] 'Sub Main' is declared more than once in 'project0'
Imports System
Imports System.Windows.Forms
Imports System.Threading
Imports System.Threading.Tasks

Namespace Tests.TestCases

  Class Program_00
        ' Error@+1 [BC30002] Attribute doesn't exist
        <Nonsense>
        Shared Sub Main() ' Noncompliant
        End Sub
    End Class

  Public Class Program_01
    Shared Sub Main() ' Noncompliant {{Add the 'STAThread' attribute to this entry point.}}
'              ^^^^
    End Sub
  End Class

  Public Class Program_02
    Public Shared Sub Main() ' Noncompliant {{Add the 'STAThread' attribute to this entry point.}}
    End Sub
  End Class

  class LowerCaseProgram
    public shared sub main(args as string()) ' Noncompliant {{Add the 'STAThread' attribute to this entry point.}}
    end sub
  end class

  Class Program_04
    <MTAThread()> Shared Sub Main() ' Noncompliant {{Change the 'MTAThread' attribute of this entry point to 'STAThread'.}}
    End Sub
  End Class

  Class Program_05
    <MTAThread()> Shared Sub Main(args As String()) ' Noncompliant {{Change the 'MTAThread' attribute of this entry point to 'STAThread'.}}
    End Sub
  End Class

  Class Program_06
    Sub Main(args As String()) ' not static method
    End Sub
  End Class

  Public Class Program_07
    <STAThread()> Shared Sub Main()
    End Sub
  End Class

  Public Class Program_08
    <STAThread> Shared Sub Main(args As String())
    End Sub
  End Class

  Class Program_09
        ' Error@+1 [BC30057] Invalid code
        <STAThread("random", 1)>
        Shared Sub Main()
        End Sub
    End Class

  Class Program_10
    <MTAThread>
    Shared Sub Pain()
    End Sub
  End Class

  Class Program_11
    Public Shared Async Function Main() As Task
      Await Task.CompletedTask
    End Function
  End Class

  Class Program_12
    <STAThread>
    Public Shared Async Function Main() As Task
        Await Task.CompletedTask
    End Function
  End Class

  Class Program_13
    <MTAThread>
    Public Shared Async Function Main() As Task
        Await Task.CompletedTask
    End Function
  End Class

End Namespace
